pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java

```diff
@@ -19,255 +19,409 @@
+  public SuccessResponse reloadSegmentDeprecated1(
       @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
-      @ApiParam(value = "realtime|offline") @QueryParam("type") String tableTypeStr) {
-    CommonConstants.Helix.TableType tableType = Constants.validateTableType(tableTypeStr);
-    return reloadAllSegmentsForTable(tableName, tableType);
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr) {
+    return reloadSegment(tableName, segmentName, tableTypeStr);
   }
 
   @Deprecated
   @GET
   @Path("tables/{tableName}/segments/{segmentName}/reload")
   @Produces(MediaType.APPLICATION_JSON)
-  @ApiOperation(value = "Reloads one segment", notes = "Reloads one segment")
-  public SuccessResponse reloadOneSegmentDeprecated(
+  @ApiOperation(value = "Reload a segment (deprecated, use 'POST /segments/{tableName}/{segmentName}/reload' instead)", notes = "Reload a segment (deprecated, use 'POST /segments/{tableName}/{segmentName}/reload' instead)")
+  public SuccessResponse reloadSegmentDeprecated2(
       @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
-      @ApiParam(value = "Name of segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
-      @ApiParam(value = "realtime|offline") @QueryParam("type") String tableTypeStr) {
-    segmentName = URIUtils.decode(segmentName);
-    CommonConstants.Helix.TableType tableType = Constants.validateTableType(tableTypeStr);
-    return reloadSegmentForTable(tableName, segmentName, tableType);
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr) {
+    return reloadSegment(tableName, segmentName, tableTypeStr);
+  }
+
+  @POST
+  @Path("segments/{tableName}/reload")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Reload all segments", notes = "Reload all segments")
+  public SuccessResponse reloadAllSegments(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr) {
+    List<String> tableNamesWithType =
+        getExistingTableNamesWithType(tableName, Constants.validateTableType(tableTypeStr));
+    int numReloadMessagesSent = 0;
+    for (String tableNameWithType : tableNamesWithType) {
+      numReloadMessagesSent += _pinotHelixResourceManager.reloadAllSegments(tableNameWithType);
+    }
+    return new SuccessResponse("Sent " + numReloadMessagesSent + " reload messages");
```

Comment:
Will be good if we can separate the number of messages between offline and realtime in the response.

Reply:
Done